modules/guides/src/GuidesStorage.php

```diff
@@ -0,0 +1,126 @@
+<?php
+
+namespace Drupal\guides;
+
+/**
+ * Copyright (C) 2019 PRONOVIX GROUP BVBA.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301,
+ * USA.
+ */
+
+use Drupal\Core\Cache\CacheBackendInterface;
+use Drupal\Core\Url;
+use Drupal\Core\Link;
+use Drupal\Core\Site\Settings;
+
+/**
+ * Class GuidesStorage.
+ */
+class GuidesStorage implements GuidesInterface {
+
+  /**
+   * The settings of the site.
+   *
+   * @var \Drupal\Core\Site\Settings
+   */
+  protected $settings;
+
+  /**
+   * The cache backend service.
+   *
+   * @var \Drupal\Core\Cache\CacheBackendInterface
+   */
+  protected $cacheBackend;
+
+  /**
+   * GuidesStorage constructor.
+   *
+   * @param \Drupal\Core\Site\Settings $settings
+   *   The settings of the site.
+   */
+  public function __construct(Settings $settings, CacheBackendInterface $cache_backend) {
+    $this->settings = $settings;
+    $this->cacheBackend = $cache_backend;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getDirectory(): string {
+    return DRUPAL_ROOT . ($this->settings->get('guides_dir') ?? '/guides');
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getDirectories(): array {
+    $directories = [];
+    $guides_directory = $this->getDirectory();
+
+    foreach (array_diff(scandir($guides_directory), ['..', '.']) as $guides_dir) {
```

Comment:
Or maybe we do not even need this `getDirectories()` function at all because the `getFiles()` method could load all markdown files directly from the parent guides dir (returned by `getDirectory()`), it does not need the subdirectories for that. You can use the https://www.php.net/manual/en/class.recursivefilteriterator.php for this approach.

Reply:
I used `RecursiveFilterIterator` but meanwhile I found a simpler solution to get all the files: I can get them by looking for files matching the pattern: `glob($directory . '/*/*.md')` but I don't know whether it is a good solution or not. And I had to use this also with the `RecursiveFilterIterator`, because the `Iterator` gives back all the files, but I don't need those files which are directly in the guides folder, so I used this `glob()` function to filter out the files with the wrong path pattern, and I realised that it gives back exactly what we need here. So the question is that is it good as it is now with the `Iterator`, or would be enough to use only the `glob()` function?

Reply:
My gut says `glob()` should be enough itself – but I'm not sure if I should listen to my gut in this case.